mir_eval/alignment.py

```diff
@@ -0,0 +1,300 @@
+        "Expected largest reference timestamp "
+        f"{np.max(reference_timestamps)} to not be "
+        f"larger than duration {duration}"
+    )
+    assert np.max(estimated_timestamps) <= duration, (
+        "Expected largest estimate timestamp "
+        f"{np.max(estimated_timestamps)} to not be "
+        f"larger than duration {duration}"
+    )
+
+    ref_starts = np.concatenate([[0], reference_timestamps])
+    ref_ends = np.concatenate([reference_timestamps, [duration]])
+    est_starts = np.concatenate([[0], estimated_timestamps])
+    est_ends = np.concatenate([estimated_timestamps, [duration]])
+    overlap_starts = np.maximum(ref_starts, est_starts)
+    overlap_ends = np.minimum(ref_ends, est_ends)
+    overlap_duration = np.sum(np.maximum(overlap_ends - overlap_starts, 0))
+    return overlap_duration / duration
+
+
+def perceptual_synchrony(offset):
+    """function perceptually weighting the deviations of onsets"""
+    skewness = 1.12244251
+    localisation = -0.22270315
+    scale = 0.29779424
+    normalisation_factor = 1.6857
+    return (1.0 / normalisation_factor) * skewnorm.pdf(
+        offset, skewness, loc=localisation, scale=scale
+    )
+
+
+def perceptual_metric(reference_timestamps, estimated_timestamps):
+    """metric based on human synchronicity perception as measured in the paper
+
+    "User-centered evaluation of lyrics to audio alignment",
+    N. Lizé-Masclef, A. Vaglio, M. Moussallam, ISMIR 2021
+
+    The parameters of this function were tuned on data collected through a user Karaoke-like
+    experiment
+    It reflects human judgment of how "synchronous" lyrics and audio stimuli are perceived
+    in that setup.
+    Beware that this metric is non-symmetrical and by construction it is also not equal to 1 at 0.
+
+    Examples
+    --------
+    >>> reference_timestamps = mir_eval.io.load_events('reference.txt')
+    >>> estimated_timestamps = mir_eval.io.load_events('estimated.txt')
+    >>> perceptual = mir_eval.align.perceptual_metric(reference_onsets, estimated_timestamps)
+
+    Parameters
+    ----------
+    reference_timestamps : np.ndarray
+        reference timestamps, in seconds
+    estimated_timestamps : np.ndarray
+        estimated timestamps, in seconds
+    Returns
+    -------
+    perceptual_score : float
+        Perceptual score, averaged over all timestamps
+    """
+    validate(reference_timestamps, estimated_timestamps)
+    perceptual_score = perceptual_synchrony(estimated_timestamps - reference_timestamps)
+    return np.mean(perceptual_score)
+
+
+def evaluate(reference_timestamps, estimated_timestamps, duration: float, **kwargs):
+    """Compute all metrics for the given reference and estimated annotations.
+    Examples
+    --------
+    >>> reference_timestamps = mir_eval.io.load_events('reference.txt')
+    >>> estimated_timestamps = mir_eval.io.load_events('estimated.txt')
+    >>> duration = max(np.max(reference_timestamps), np.max(estimated_timestamps)) + 10
```

Comment:
Where does this 10 come from? Is this really standard?

Reply:
10 is just an arbitrary number here. the duration should be set to the total duration of the audio, so the PCS metric can be correctly computed. should i change this to something like
`audio, sr = librosa.load(audio_path, mono=True)
duration = float(len(audio)) / float(sr)`

or simply mention in the description above that duration needs to be the audio duration?

Reply:
no, librosa should not be a dependency of mir_eval.  We usually infer duration from the reference annotation so that it can be consistently applied across all estimates.

Reply:
Ah, maybe this is a misunderstanding but I suggest this only as example code in the doc part of the function, not as actual implementation. If we want to keep librosa completely out of this, the example code in the doc of this function could also be
```
    >>> reference_timestamps = mir_eval.io.load_events('reference.txt')
    >>> estimated_timestamps = mir_eval.io.load_events('estimated.txt')
    >>> pcs = mir_eval.align.pcs(reference_onsets, estimated_timestamps, duration)
```
with the hint that duration needs to be provided by the user and indicates the total song length.

Btw, the reason why I am not including the duration as another timestamp in the `reference_timestamps` and `estimated_timestamps` is because it creates logic branching in most of the other metrics, where you would then want to ignore it, since it's technically not a timestamp that needs to be predicted. for example, for the percentage of correctly detected timestamps, it doesn't make sense to include the end of the song as a timestamp to be predicted. PCS is a bit special in this sense, for computing the segment overlap you need the end position of the last segment, which is the end of the song.

So hopefully this clears up some things about the metrics. Do you agree that changing the example code in the docs like i suggested is the best way to go about this?

Reply:
I don't know if this is your problem to solve, but it's problematic to me that the annotation format for this task is incomplete. If the song duration is necessary to perform evaluation, then the annotation format should include the duration. This is definitely a red flag for evaluation inconsistencies in the future - let's say you load your audio files with `librosa` to get the duration; I load them with some other package/binary that always appends 1s of silence (sillier things have happened). Now our scores won't be comparable, even though we are using the same annotations.

How is this handled in MIREX and in any other packages? I feel either the duration should be included in the annotation, or there should be some convention (e.g. duration = 10s after last event, though this seems arbitrary and weird).

Reply:
We generally err towards community consensus with `mir_eval`. Consensus can be implicit, based on annotation formats. Are there a large number of annotations of songs available in the "1D list"-style? In that case, and given that MIREX doesn't use a duration, I would not suggest we create a new annotation format (and generally agree that it's more work than is warranted). WDYT of this solution: make the duration parameter optional (default to None); when not passed, do the MIREX-style evaluation; when passed, do what you are currently doing. Then add a note mentioning the MIREX convention and the "right" way and leave it up to users. In the evaluate docstring, you can just not pass in a duration, etc.

Reply:
Yes, I would say at least for lyrics alignment, mostly the annotation files are also in this 1D array format, and usually don't specify the audio duration in it. 
I think that solution sounds good! This way the `evaluate` metric also runs out of the box without needing the duration. And then the community can still switch to the other version of the metric if it deems that one really as better, but we don't need to decide that ourselves.
Will implement this now!

Reply:
👍